### PR TITLE
fix(api): Correct format for fixed-layout epub when using Kobo Sync

### LINKED ISSUE #3151: [Bug] Kobo Sync using wrong format for fixed-layout epubs
### What Went Wrong?

When using Kobo Sync with fixed-layout epubs (for example comics), the format is specified as EPUB3 (or KEPUB) instead of EPUB3FL, which causes the ereader to use the wrong epub viewer.
A similar issue used to happen with calibre-web: https://github.com/janeczku/calibre-web/pull/2262

The normal fixed layout viewer hides header and footer to show full page images, and shows two pages in landscape mode.
However, as it was not set it instead showed the default viewer, which shrinks the images and only shows one page in landscape.

### How Can We Reproduce This?

1. Get a fixed-layout epub (I dont have any without copyright protection, if requested I'll make a reproducer)
2. Upload to booklore, enable kobo sync
3. Do a sync on the ereader, and you'll notice the images not taking up the whole screen, and only showing one page in landscape

### What Should Have Happened?

It should open the correct fixed layout viewer on the ereader.

### Screenshots or Error Messages (Optional)

A user in discord had the same issue, once / if I get their permission I'll upload comparisons here.
Or I'll do it once I'm home.

### Any Ideas on How to Fix This? (Optional)

I already fixed the issue, and am creating the issue for permission to open a PR.

Essentially porting over the same change from calibre-web.

### Your Setup

- Booklore version: Issue reproduced on 2.0.4 and develop at 5e133779.
- OS: Arch Linux
- Installation: Docker Compose
- Browser: Firefox 148
- Storage Type: Tested on SSD and HDD

### Before Submitting

- [x] I've searched existing issues and confirmed this bug hasn't been reported yet

### PR DESCRIPTION
**Pull request by N00byKing**
## 📝 Description

This PR adds detection for the `rendition:layout` property in fixed-layout epubs.
For those where it is set to `pre-paginated` EPUB3FL is used as the format instead of KEPUB/EPUB3,
making the ereader use the comic viewer instead of the default one, making pages take up the entirety of the screen.

Essentially a port of https://github.com/janeczku/calibre-web/pull/2262

**Linked Issue:** Fixes #3151 

> **Required.** Every PR must reference an approved issue. If no issue exists, [open one](https://github.com/booklore-app/booklore/issues/new) and wait for maintainer approval before submitting a PR. Unsolicited PRs without a linked issue will be closed.

## 🏷️ Type of Change

- [x] Bug fix
- [ ] New feature
- [ ] Enhancement to existing feature
- [ ] Refactor (no behavior change)
- [ ] Breaking change (existing functionality affected)
- [ ] Documentation update

## 🔧 Changes

On sync, epubs are inspected for the `rendition:layout` property.
For that, I copied over the extraction from `EpubMetadataExtractor.java` and modified it to suit my needs.
Not very happy about that, but didnt find a much cleaner way as I am quite unfamiliar with the codebase.
Let me know if there's a better way to do it.

If its set to `pre-paginated` the format is set accordingly.
NOTE: Should this also apply to converted cbx files? I really dont know.
They *are* comics, but as I'm completely unfamiliar with them I left it as is to avoid false positives;
but it'd be trivial to apply to those as well.

## 🧪 Testing (MANDATORY)

> **PRs without this section filled out will be closed.** "Tests pass" or "Tested locally" is not sufficient. You must provide specifics.

**Manual testing steps you performed:**

1. Manually inspected the JSON output from the /v1/library/sync endpoint
2. Synced with my actual device with the reproducer from the issue

**Regression testing:**

- Tested using a normal epub as well

**Edge cases covered:**

- Dont know of any

**Test output:**

<details>
<summary>Backend test output (<code>./gradlew test</code>)</summary>

```
[archuser@doggo booklore-api]$ ./gradlew test

> Task :compileJava
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Management of bidirectional association persistent attributes is deprecated and will be removed. Set the value to 'false' to get rid of this warning
OpenJDK 64-Bit Server VM warning: Sharing is only supported for boot loader classes because bootstrap classpath has been appended
WARNING: A terminally deprecated method in sun.misc.Unsafe has been called
WARNING: sun.misc.Unsafe::objectFieldOffset has been called by net.bytebuddy.dynamic.loading.ClassInjector$UsingUnsafe$Dispatcher$CreationAction (file:/home/archuser/.gradle/caches/modules-2/files-2.1/net.bytebuddy/byte-buddy/1.17.8/af5735f63d00ca47a9375fae5c7471a36331c6ed/byte-buddy-1.17.8.jar)
WARNING: Please consider reporting this to the maintainers of class net.bytebuddy.dynamic.loading.ClassInjector$UsingUnsafe$Dispatcher$CreationAction
WARNING: sun.misc.Unsafe::objectFieldOffset will be removed in a future release
2026-03-02T19:45:25.200+01:00  INFO 147971 --- [booklore-api] [ionShutdownHook] o.s.m.s.b.SimpleBrokerMessageHandler     : Stopping...
2026-03-02T19:45:25.201+01:00  INFO 147971 --- [booklore-api] [ionShutdownHook] o.s.m.s.b.SimpleBrokerMessageHandler     : BrokerAvailabilityEvent[available=false, SimpleBrokerMessageHandler [org.springframework.messaging.simp.broker.DefaultSubscriptionRegistry@5483b79a]]
2026-03-02T19:45:25.201+01:00  INFO 147971 --- [booklore-api] [ionShutdownHook] o.s.m.s.b.SimpleBrokerMessageHandler     : Stopped.
2026-03-02T19:45:25.210+01:00  INFO 147971 --- [booklore-api] [opFolderWatcher] o.b.s.b.BookdropMonitoringService        : Bookdrop monitor thread interrupted
2026-03-02T19:45:25.210+01:00  INFO 147971 --- [booklore-api] [ionShutdownHook] o.b.s.b.BookdropMonitoringService        : Stopped bookdrop folder monitor
2026-03-02T19:45:25.723+01:00  INFO 147971 --- [booklore-api] [opFileProcessor] o.b.s.b.BookdropEventHandlerService      : File processing thread interrupted, shutting down.
2026-03-02T19:45:25.724+01:00  INFO 147971 --- [booklore-api] [ionShutdownHook] o.b.s.monitoring.MonitoringService       : Shutting down monitoring service...
2026-03-02T19:45:25.725+01:00  INFO 147971 --- [booklore-api] [ionShutdownHook] o.b.s.watcher.LibraryFileEventProcessor  : Shutting down LibraryFileEventProcessor...
2026-03-02T19:45:25.725+01:00  WARN 147971 --- [booklore-api] [        async-0] o.b.service.monitoring.MonitoringTask    : WatchService has been closed. Stopping monitoring.
2026-03-02T19:45:25.729+01:00  INFO 147971 --- [booklore-api] [ionShutdownHook] j.LocalContainerEntityManagerFactoryBean : Closing JPA EntityManagerFactory for persistence unit 'default'
2026-03-02T19:45:25.745+01:00  INFO 147971 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown initiated...
2026-03-02T19:45:25.745+01:00  INFO 147971 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown completed.
2026-03-02T19:45:25.750+01:00  INFO 147971 --- [booklore-api] [opFolderWatcher] o.b.s.b.BookdropMonitoringService        : Bookdrop monitor thread interrupted
2026-03-02T19:45:25.750+01:00  INFO 147971 --- [booklore-api] [ionShutdownHook] o.b.s.b.BookdropMonitoringService        : Stopped bookdrop folder monitor
2026-03-02T19:45:25.750+01:00  INFO 147971 --- [booklore-api] [opFileProcessor] o.b.s.b.BookdropEventHandlerService      : File processing thread interrupted, shutting down.
2026-03-02T19:45:25.751+01:00  INFO 147971 --- [booklore-api] [ionShutdownHook] o.b.s.monitoring.MonitoringService       : Shutting down monitoring service...
2026-03-02T19:45:25.751+01:00  INFO 147971 --- [booklore-api] [ionShutdownHook] o.b.s.watcher.LibraryFileEventProcessor  : Shutting down LibraryFileEventProcessor...
2026-03-02T19:45:25.751+01:00  WARN 147971 --- [booklore-api] [        async-0] o.b.service.monitoring.MonitoringTask    : WatchService has been closed. Stopping monitoring.
2026-03-02T19:45:25.756+01:00  INFO 147971 --- [booklore-api] [ionShutdownHook] j.LocalContainerEntityManagerFactoryBean : Closing JPA EntityManagerFactory for persistence unit 'default'
2026-03-02T19:45:25.758+01:00  INFO 147971 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown initiated...
2026-03-02T19:45:25.758+01:00  INFO 147971 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown completed.
2026-03-02T19:45:25.763+01:00  INFO 147971 --- [booklore-api] [opFolderWatcher] o.b.s.b.BookdropMonitoringService        : Bookdrop monitor thread interrupted
2026-03-02T19:45:25.763+01:00  INFO 147971 --- [booklore-api] [ionShutdownHook] o.b.s.b.BookdropMonitoringService        : Stopped bookdrop folder monitor
2026-03-02T19:45:25.763+01:00  INFO 147971 --- [booklore-api] [opFileProcessor] o.b.s.b.BookdropEventHandlerService      : File processing thread interrupted, shutting down.
2026-03-02T19:45:25.764+01:00  INFO 147971 --- [booklore-api] [ionShutdownHook] o.b.s.monitoring.MonitoringService       : Shutting down monitoring service...
2026-03-02T19:45:25.764+01:00  INFO 147971 --- [booklore-api] [ionShutdownHook] o.b.s.watcher.LibraryFileEventProcessor  : Shutting down LibraryFileEventProcessor...
2026-03-02T19:45:25.765+01:00  WARN 147971 --- [booklore-api] [        async-0] o.b.service.monitoring.MonitoringTask    : WatchService has been closed. Stopping monitoring.
2026-03-02T19:45:25.773+01:00  INFO 147971 --- [booklore-api] [ionShutdownHook] j.LocalContainerEntityManagerFactoryBean : Closing JPA EntityManagerFactory for persistence unit 'default'
2026-03-02T19:45:25.782+01:00  INFO 147971 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown initiated...
2026-03-02T19:45:25.782+01:00  INFO 147971 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown completed.
2026-03-02T19:45:25.788+01:00  INFO 147971 --- [booklore-api] [opFolderWatcher] o.b.s.b.BookdropMonitoringService        : Bookdrop monitor thread interrupted
2026-03-02T19:45:25.788+01:00  INFO 147971 --- [booklore-api] [ionShutdownHook] o.b.s.b.BookdropMonitoringService        : Stopped bookdrop folder monitor
2026-03-02T19:45:25.788+01:00  INFO 147971 --- [booklore-api] [opFileProcessor] o.b.s.b.BookdropEventHandlerService      : File processing thread interrupted, shutting down.
2026-03-02T19:45:25.789+01:00  INFO 147971 --- [booklore-api] [ionShutdownHook] o.b.s.monitoring.MonitoringService       : Shutting down monitoring service...
2026-03-02T19:45:25.790+01:00  INFO 147971 --- [booklore-api] [ionShutdownHook] o.b.s.watcher.LibraryFileEventProcessor  : Shutting down LibraryFileEventProcessor...
2026-03-02T19:45:25.790+01:00  WARN 147971 --- [booklore-api] [        async-0] o.b.service.monitoring.MonitoringTask    : WatchService has been closed. Stopping monitoring.
2026-03-02T19:45:25.796+01:00  INFO 147971 --- [booklore-api] [ionShutdownHook] j.LocalContainerEntityManagerFactoryBean : Closing JPA EntityManagerFactory for persistence unit 'default'
2026-03-02T19:45:25.804+01:00  INFO 147971 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown initiated...
2026-03-02T19:45:25.805+01:00  INFO 147971 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown completed.
2026-03-02T19:45:25.809+01:00  INFO 147971 --- [booklore-api] [opFolderWatcher] o.b.s.b.BookdropMonitoringService        : Bookdrop monitor thread interrupted
2026-03-02T19:45:25.809+01:00  INFO 147971 --- [booklore-api] [ionShutdownHook] o.b.s.b.BookdropMonitoringService        : Stopped bookdrop folder monitor
2026-03-02T19:45:25.809+01:00  INFO 147971 --- [booklore-api] [opFileProcessor] o.b.s.b.BookdropEventHandlerService      : File processing thread interrupted, shutting down.
2026-03-02T19:45:25.810+01:00  INFO 147971 --- [booklore-api] [ionShutdownHook] o.b.s.monitoring.MonitoringService       : Shutting down monitoring service...
2026-03-02T19:45:25.811+01:00  WARN 147971 --- [booklore-api] [        async-0] o.b.service.monitoring.MonitoringTask    : WatchService has been closed. Stopping monitoring.
2026-03-02T19:45:25.811+01:00  INFO 147971 --- [booklore-api] [ionShutdownHook] o.b.s.watcher.LibraryFileEventProcessor  : Shutting down LibraryFileEventProcessor...
2026-03-02T19:45:25.819+01:00  INFO 147971 --- [booklore-api] [ionShutdownHook] j.LocalContainerEntityManagerFactoryBean : Closing JPA EntityManagerFactory for persistence unit 'default'
2026-03-02T19:45:25.826+01:00  INFO 147971 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown initiated...
2026-03-02T19:45:25.827+01:00  INFO 147971 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown completed.
2026-03-02T19:45:25.832+01:00  INFO 147971 --- [booklore-api] [opFolderWatcher] o.b.s.b.BookdropMonitoringService        : Bookdrop monitor thread interrupted
2026-03-02T19:45:25.832+01:00  INFO 147971 --- [booklore-api] [ionShutdownHook] o.b.s.b.BookdropMonitoringService        : Stopped bookdrop folder monitor
2026-03-02T19:45:25.832+01:00  INFO 147971 --- [booklore-api] [opFileProcessor] o.b.s.b.BookdropEventHandlerService      : File processing thread interrupted, shutting down.
2026-03-02T19:45:25.833+01:00  INFO 147971 --- [booklore-api] [ionShutdownHook] o.b.s.monitoring.MonitoringService       : Shutting down monitoring service...
2026-03-02T19:45:25.834+01:00  INFO 147971 --- [booklore-api] [ionShutdownHook] o.b.s.watcher.LibraryFileEventProcessor  : Shutting down LibraryFileEventProcessor...
2026-03-02T19:45:25.834+01:00  WARN 147971 --- [booklore-api] [        async-0] o.b.service.monitoring.MonitoringTask    : WatchService has been closed. Stopping monitoring.
2026-03-02T19:45:25.840+01:00  INFO 147971 --- [booklore-api] [ionShutdownHook] j.LocalContainerEntityManagerFactoryBean : Closing JPA EntityManagerFactory for persistence unit 'default'
2026-03-02T19:45:25.848+01:00  INFO 147971 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown initiated...
2026-03-02T19:45:25.849+01:00  INFO 147971 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown completed.

[Incubating] Problems report is available at: file:///home/archuser/Desktop/booklore/booklore-api/build/reports/problems/problems-report.html

BUILD SUCCESSFUL in 1m 6s
6 actionable tasks: 3 executed, 3 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.3.1/userguide/configuration_cache_enabling.html
```

</details>

<details>
<summary>Frontend test output (<code>ng test</code>)</summary>

```
root@49429b1c21da:/home# ./node_modules/.bin/ng test
Using Vitest configuration file: /home/vitest-base.config.ts
Initial chunk files                                                            | Names                                                                       |  Raw size
spec-app-features-magic-shelf-component-magic-shelf-component.js               | spec-app-features-magic-shelf-component-magic-shelf-component               |   1.09 MB |
spec-app-features-magic-shelf-service-book-rule-evaluator.service.js           | spec-app-features-magic-shelf-service-book-rule-evaluator.service           |  65.08 kB |
spec-app-app.component.js                                                      | spec-app-app.component                                                      |  58.99 kB |
chunk-J7OUGXQA.js                                                              | -                                                                           |  53.30 kB |
styles.css                                                                     | styles                                                                      |  30.99 kB |
chunk-6GQKDBEE.js                                                              | -                                                                           |  25.09 kB |
spec-app-features-magic-shelf-service-book-rule-evaluator-metadata-presence.js | spec-app-features-magic-shelf-service-book-rule-evaluator-metadata-presence |  23.21 kB |
chunk-ZRYX4BS5.js                                                              | -                                                                           |  15.97 kB |
spec-app-features-magic-shelf-service-magic-shelf-utils.js                     | spec-app-features-magic-shelf-service-magic-shelf-utils                     |  11.36 kB |
spec-app-core-security-auth-initializer.js                                     | spec-app-core-security-auth-initializer                                     |   8.69 kB |
chunk-PUIDBMXW.js                                                              | -                                                                           |   6.10 kB |
chunk-Y7BNZHZM.js                                                              | -                                                                           |   4.86 kB |
spec-app-features-book-service-library-health.service.js                       | spec-app-features-book-service-library-health.service                       |   2.95 kB |
chunk-3J3QIHZF.js                                                              | -                                                                           |   2.19 kB |
chunk-2UARJPKY.js                                                              | -                                                                           |   2.07 kB |
chunk-HPNCK62B.js                                                              | -                                                                           |   1.33 kB |
init-testbed.js                                                                | init-testbed                                                                |   1.27 kB |
vitest-mock-patch.js                                                           | vitest-mock-patch                                                           | 481 bytes |
spec-app-app.js                                                                | spec-app-app                                                                | 202 bytes |
polyfills.js                                                                   | polyfills                                                                   | 121 bytes |

                                                                               | Initial total                                                               |   1.41 MB

Application bundle generation complete. [4.085 seconds] - 2026-03-02T18:59:33.632Z

Watch mode enabled. Watching for file changes...

 DEV  v4.0.18 /home

 ✓  booklore  src/app/app.spec.ts (1 test) 3ms
 ✓  booklore  src/app/features/magic-shelf/service/magic-shelf-utils.spec.ts (46 tests) 17ms
 ✓  booklore  src/app/features/magic-shelf/service/book-rule-evaluator.service.spec.ts (199 tests) 231ms
 ✓  booklore  src/app/features/magic-shelf/service/book-rule-evaluator-metadata-presence.spec.ts (85 tests) 126ms
 ✓  booklore  src/app/core/security/auth-initializer.spec.ts (2 tests) 19ms
 ✓  booklore  src/app/features/book/service/library-health.service.spec.ts (6 tests) 28ms
 ✓  booklore  src/app/app.component.spec.ts (7 tests) 219ms
 ✓  booklore  src/app/features/magic-shelf/component/magic-shelf-component.spec.ts (73 tests) 362ms

 Test Files  8 passed (8)
      Tests  419 passed (419)
   Start at  18:59:33
   Duration  2.60s (transform 1.03s, setup 2.13s, import 2.91s, tests 1.00s, environment 3.53s)

JUNIT report written to /home/test-results/vitest-results.xml
 PASS  Waiting for file changes...
       press h to show help, press q to quit
```

</details>

## 📸 Screen Recording / Screenshots (MANDATORY)

> Every PR must include a **screen recording or screenshots** showing the change working end-to-end in a running local instance (both backend and frontend). This means you must have actually built, run, and tested the code yourself. PRs without visual proof will be closed without review.

Single Page full-screen of reproducer file: 
Double Page in landscape: 

---

## ✅ Pre-Submission Checklist

> **All boxes must be checked before requesting review.** Incomplete PRs will be closed without review. No exceptions.

- [x] This PR is linked to an approved issue
- [x] Code follows project [backend and frontend conventions](../CONTRIBUTING.md#backend-conventions)
- [x] Branch is up to date with `develop` (merge conflicts resolved)
- [x] I ran the full stack locally (backend + frontend + database) and verified the change works
- [ ] Automated tests added or updated to cover changes (backend **and** frontend)
- [x] All tests pass locally and output is pasted above
- [x] Screen recording or screenshots are attached above proving the change works
- [x] PR is a single focused change (one bug fix OR one feature, not multiple unrelated changes)
- [x] PR is reasonably scoped (PRs over 1000+ changed lines will be closed, split into smaller PRs)
- [x] No unsolicited refactors, cleanups, or "improvements" are bundled in
- [x] Flyway migration versioning is correct _(if schema was modified)_
- [x] Documentation PR submitted to [booklore-docs](https://github.com/booklore-app/booklore-docs) _(if user-facing changes)_

### 🤖 AI-Assisted Contributions

> **If any part of this PR was generated or assisted by AI tools (Copilot, Claude, ChatGPT, etc.), all items below are mandatory.** You are fully responsible for every line you submit. "The AI wrote it" is not an excuse, and AI-generated PRs that clearly haven't been reviewed are the #1 reason PRs get closed.

PR was not AI assisted.

---

## 💬 Additional Context _(optional)_

Did not add a test for it as it would require an epub file to be added for it specifically.
Since that has not happened yet for other tests, I havent added the reproducer / written a test for it.

### CODE REVIEW

File: booklore-api/src/main/java/org/booklore/service/kobo/KoboEntitlementService.java [hunk 1]
```diff
@@ -279,6 +292,46 @@ public KoboBookMetadata getMetadataForBook(long bookId, String token) {
...
     }
 
+    private boolean checkEpubFixedLayout(BookEntity book) {
+        if (book.getPrimaryBookFile().getBookType() != BookFileType.EPUB) return false;
+        Path filePath = book.getFullFilePath();
+        if (filePath == null) return false;
+
+        File epubFile = book.getFullFilePath().toFile();
+        try (ZipFile zip = new ZipFile(epubFile)) {
+            DocumentBuilder builder = SecureXmlUtils.createSecureDocumentBuilder(true);
```
Comment (acx10) [inline booklore-api/src/main/java/org/booklore/service/kobo/KoboEntitlementService.java]: This duplicates the container.xml → OPF parsing that already exists in `EpubMetadataExtractor.extractMetadata()` and `EpubReaderService.parseContainerXml()`. Could we extract a small shared helper (e.g. in `EpubMetadataExtractor` or a new `EpubUtils`) that opens an EPUB, finds the OPF, and returns the parsed `Document`? Then this method just checks the `<meta property="rendition:layout">` on the returned doc. Would keep things DRY and easier to maintain.

File: booklore-api/src/main/java/org/booklore/service/kobo/KoboEntitlementService.java [hunk 2]
```diff
@@ -279,6 +292,46 @@ public KoboBookMetadata getMetadataForBook(long bookId, String token) {
         return mapToKoboMetadata(books.getFirst(), token);
     }
 
+    private boolean checkEpubFixedLayout(BookEntity book) {
```
Comment (acx10) [inline booklore-api/src/main/java/org/booklore/service/kobo/KoboEntitlementService.java]: This opens and parses the EPUB zip on every sync request. For users with large libraries (10k+ books), that's potentially thousands of zip files opened during a single sync. Since `rendition:layout` doesn't change after import, it would be better to detect this once during metadata extraction and store it (e.g. as a field on `BookMetadataEntity` or `BookFileEntity`), then just read from the DB during sync.

File: booklore-api/src/main/java/org/booklore/service/kobo/KoboEntitlementService.java [hunk 3]
```diff
@@ -279,6 +292,46 @@ public KoboBookMetadata getMetadataForBook(long bookId, String token) {
...
+                if (opfHdr == null) return false;
+
+                try (InputStream in = zip.getInputStream(opfHdr)) {
+                    Document doc = builder.parse(in);
+                    NodeList manifestItems = doc.getElementsByTagName("meta");
+
+                    for (int i = 0; i < manifestItems.getLength(); i++) {
+                        Element item = (Element) manifestItems.item(i);
+                        String prop = item.getAttribute("property");
+                        if (prop.equals("rendition:layout") && item.getTextContent().equals("pre-paginated")) return true;
```
Comment (acx10) [inline booklore-api/src/main/java/org/booklore/service/kobo/KoboEntitlementService.java]: Some EPUBs use the older EPUB 2 `<meta name="..." content="...">` syntax instead of the EPUB 3 `<meta property="...">` syntax. Might be worth also checking `item.getAttribute("name").equals("rendition:layout") && item.getAttribute("content").equals("pre-paginated")` to handle both formats.

Comment (N00byKing) [reply booklore-api/src/main/java/org/booklore/service/kobo/KoboEntitlementService.java]: Fixed layout was only introduced with epub3, is that still relevant?

File: booklore-api/src/main/java/org/booklore/service/kobo/KoboEntitlementService.java (hunk 1 above)
Comment (N00byKing) [reply booklore-api/src/main/java/org/booklore/service/kobo/KoboEntitlementService.java]: I've merged most places that use the opf file by adding a few public helpers in epubreaderservice.
It seemed like the best option considering it already has a few xml helpers in place.

I've also rerun the backend tests and they still pass, and checking with my ereader the syncing also still works.

File: booklore-api/src/main/java/org/booklore/service/kobo/KoboEntitlementService.java (hunk 2 above)
Comment (N00byKing) [reply booklore-api/src/main/java/org/booklore/service/kobo/KoboEntitlementService.java]: Since this is a epub-specific property I guess I can put it in `BookFileEntity`.
Will see about that tomorrow.

Comment (N00byKing) [reply booklore-api/src/main/java/org/booklore/service/kobo/KoboEntitlementService.java]: I'm now caching the fixed layout metadata in `BookFileEntity`, and added a database migration for it.
Reran the backend tests and manually inspecteded the /v1/library/sync endpoint output (Don't have my ereader at hand atm), everything looks good so far.

**Comment by N00byKing** [comment]
Just a note, the diff is f'ed up because of the indentation changes after the code dedup.
If you inspect the changes on a better diff viewer (for example in vscode) most of it is just removing two levels of indentation.

File: booklore-api/src/main/java/org/booklore/service/reader/EpubReaderService.java [hunk 4]
```diff
@@ -352,6 +347,17 @@ private String parseContainerXml(ZipFile zipFile) throws Exception {
         return fullPath;
     }
 
+    public static String getOPFPath(File epubFile) throws Exception {
+        ZipFile zip = new ZipFile(epubFile);
+        return parseContainerXml(zip);
+    }
```
Comment (acx10) [inline booklore-api/src/main/java/org/booklore/service/reader/EpubReaderService.java:354]: Neither method closes the `ZipFile`. This is Apache Commons Compress `ZipFile` which holds native file handles. Before this PR, `EpubMetadataExtractor` used `net.lingala.zip4j.ZipFile` with try-with-resources everywhere. This refactoring replaced all of those with calls to these two methods, so every caller now leaks. The affected call sites are `extractMetadata`, `checkEpubFixedLayout`, `findManifestCoverByHeuristic` (calls both methods, so 2 leaks), and `findCoverImageHrefInOpf` (also calls both). That's 6 leaked file handles per book import. Both methods need try-with-resources.

File: booklore-api/src/main/java/org/booklore/service/fileprocessor/EpubProcessor.java [hunk 5]
```diff
@@ -161,5 +161,11 @@ private void setBookMetadata(BookEntity bookEntity) {
                     .collect(Collectors.toSet());
             bookCreatorService.addTagsToBook(validTags, bookEntity);
         }
+
+        BookFileEntity epubFileEnt = bookEntity.getBookFiles().stream()
+                .filter(bf -> bf.getBookType() == BookFileType.EPUB && bf.isBook())
+                .findFirst()
+                .orElse(null);
+        epubFileEnt.setFixedLayout(epubMetadataExtractor.checkEpubFixedLayout(bookFile));
```
Comment (acx10) [inline booklore-api/src/main/java/org/booklore/service/fileprocessor/EpubProcessor.java]: `orElse(null)` followed by an unconditional method call. In practice this probably won't NPE since the book was just created as EPUB type, but it's still fragile. An `ifPresent` or just dropping the `orElse(null)` in favor of `.ifPresent(ent -> ent.setFixedLayout(...))` would be safer.

File: booklore-api/src/main/java/org/booklore/service/fileprocessor/EpubProcessor.java [hunk 6]
```diff
@@ -161,5 +161,11 @@ private void setBookMetadata(BookEntity bookEntity) {
                     .collect(Collectors.toSet());
             bookCreatorService.addTagsToBook(validTags, bookEntity);
         }
+
+        BookFileEntity epubFileEnt = bookEntity.getBookFiles().stream()
+                .filter(bf -> bf.getBookType() == BookFileType.EPUB && bf.isBook())
```
Comment (acx10) [inline booklore-api/src/main/java/org/booklore/service/fileprocessor/EpubProcessor.java]: `extractMetadata` (called on line 103 above) already opens and parses the full OPF document. Then `checkEpubFixedLayout` on line 169 opens and parses the same EPUB again. The `rendition:layout` check could be done inside `extractMetadata` (adding a field to `BookMetadata`) to avoid double-parsing every EPUB on import.

File: booklore-api/src/main/java/org/booklore/service/kobo/KoboEntitlementService.java [hunk 7]
```diff
@@ -328,7 +329,9 @@ private KoboBookMetadata mapToKoboMetadata(BookEntity book, String token) {
         boolean isCbxFile = primaryFile.getBookType() == BookFileType.CBX;
 
         if (koboSettings != null) {
-            if (isEpubFile && koboSettings.isConvertToKepub()) {
+            if (isEpubFile && primaryFile.isFixedLayout()) {
```
Comment (acx10) [inline booklore-api/src/main/java/org/booklore/service/kobo/KoboEntitlementService.java:332]: The fixed-layout override isn't gated by any setting. If a user has KEPUB conversion off, a fixed-layout EPUB now becomes `EPUB3FL` instead of `EPUB3`. Is that intentional? Probably yes (Kobo needs the format hint regardless), just wanted to confirm.

Comment (N00byKing) [reply booklore-api/src/main/java/org/booklore/service/kobo/KoboEntitlementService.java:332]: Yeah, thats intentional. I dont see a reason to use `EPUB3` instead of `EPUB3FL` for a pre-paginated epub, so I just made it always the case.
For all the conversions I wasn't sure, which is why it only applies to pure epubs atm.

File: booklore-api/src/main/java/org/booklore/service/reader/EpubReaderService.java (hunk 4 above)
Comment (N00byKing) [reply booklore-api/src/main/java/org/booklore/service/reader/EpubReaderService.java:354]: My bad, I didn't know about that.
Should be fixed now.

File: booklore-api/src/main/java/org/booklore/service/fileprocessor/EpubProcessor.java (hunk 6 above)
Comment (N00byKing) [reply booklore-api/src/main/java/org/booklore/service/fileprocessor/EpubProcessor.java]: Done. Since I just integrated it into the existing flow, this also resolves the epub2 thing anyway.
